File: python/src/transtex/parsers/apa.py
"""Parse APA-style citations (6th/7th simplified)."""
from __future__ import annotations

import re

from ..reference import Reference
from .common import clean_locator, generate_cite_key, normalize_pages, split_authors_delimited, strip_trailing_period


def parse_apa_citation(text: str) -> Reference:
    raw = text.strip()
    if not raw:
        raise ValueError("Empty APA citation string")

    year_match = re.search(r"\(([^)]+)\)\.", raw)
    if not year_match:
        raise ValueError("APA citation missing year segment '(year).'")

    authors_segment = raw[: year_match.start()].strip()
    remainder = raw[year_match.end() :].strip()
    year = year_match.group(1).strip() or None

    if ". " not in remainder:
        raise ValueError("APA citation missing title/container separation")
    title, remainder = remainder.split(". ", 1)
    title = strip_trailing_period(title.strip())

    locator = None
    locator_match = re.search(r"(https?://\S+|10\.\S+)$", remainder)
    if locator_match:
        locator = locator_match.group(1).strip()
        remainder = remainder[: locator_match.start()].strip()

    container_segment = remainder.rstrip(".").strip()
    container, volume, issue, pages = _parse_container(container_segment)

    authors = _parse_authors(authors_segment)
    cite_key = generate_cite_key(authors, year, title)
    reference = Reference(
        entry_type="article",
        cite_key=cite_key,
        title=title or None,
        authors=authors,
        journal=container,
        volume=volume,
        issue=issue,
        pages=normalize_pages(pages),
        year=year,
    )
    doi, url = clean_locator(locator)
    reference.doi = doi
    reference.url = url
    return reference


def _parse_container(segment: str) -> tuple[str | None, str | None, str | None, str | None]:
    if not segment:
        return None, None, None, None
    match = re.match(
        r"(?P<container>.+?)(?:,\s*(?P<volume>\d+)(?:\((?P<issue>[^)]+)\))?)?(?:,\s*(?P<pages>[\w\-–]+))?$",
        segment,
    )
    if not match:
        return segment, None, None, None
    return (
        (match.group("container") or "").strip() or None,
        (match.group("volume") or "").strip() or None,
        (match.group("issue") or "").strip() or None,
        (match.group("pages") or "").strip() or None,
    )


def _parse_authors(segment: str) -> list[str]:
    parts = split_authors_delimited(segment, separators=[", &", " &", " and "])
    authors = []
    for part in parts:
        name = part.strip().lstrip("& ").strip()
        if not name:
            continue
        if not name.endswith("."):
            name += "."
        authors.append(name)
    return authors


__all__ = ["parse_apa_citation"]

File: typescript/src/parsers/apa.ts
import { Reference } from "../reference";
import { applyLocator, generateCiteKey, normalizePages, splitAuthorsDelimiters, stripTrailingPeriod } from "./shared";

export function parseApaCitation(text: string): Reference {
    const raw = text.trim();
    if (!raw) {
        throw new Error("Empty APA citation string");
    }

    const yearMatch = raw.match(/\(([^)]+)\)\./);
    if (!yearMatch) {
        throw new Error("APA citation missing year segment '(year).'");
    }

    const authorsSegment = raw.slice(0, yearMatch.index).trim();
    const remainder = raw.slice((yearMatch.index ?? 0) + yearMatch[0].length).trim();
    const year = yearMatch[1].trim() || undefined;

    const titleSplit = remainder.split(". ");
    const title = stripTrailingPeriod(titleSplit[0].trim());
    const afterTitle = titleSplit.slice(1).join(". ").trim();

    let locator: string | undefined;
    const locatorMatch = afterTitle.match(/(https?:\/\/\S+|10\.\S+)$/);
    let containerSegment = afterTitle;
    if (locatorMatch) {
        locator = locatorMatch[1].trim();
        containerSegment = afterTitle.slice(0, locatorMatch.index).trim();
    }
    containerSegment = containerSegment.replace(/\.$/, "").trim();

    const { container, volume, issue, pages } = parseContainer(containerSegment);
    const authors = parseAuthors(authorsSegment);
    const reference = new Reference({
        entryType: container ? "article" : "book",
        citeKey: generateCiteKey(authors, year, title),
        title: title || undefined,
        authors,
        journal: container,
        publisher: !volume && !issue && !pages ? container : undefined,
        volume,
        issue,
        pages: normalizePages(pages),
        year,
    });
    applyLocator(reference, locator);
    return reference;
}

function parseContainer(segment: string): { container?: string; volume?: string; issue?: string; pages?: string } {
    if (!segment) return {};
    const match = segment.match(
        /(?<container>.+?)(?:,\s*(?<volume>\d+)(?:\((?<issue>[^)]+)\))?)?(?:,\s*(?<pages>[\w\-–]+))?$/
    );
    if (!match || !match.groups) {
        return { container: segment };
    }
    const { container, volume, issue, pages } = match.groups;
    return {
        container: container?.trim() || undefined,
        volume: volume?.trim() || undefined,
        issue: issue?.trim() || undefined,
        pages: pages?.trim() || undefined,
    };
}

function parseAuthors(segment: string): string[] {
    const parts = splitAuthorsDelimiters(segment, /\s*(?:,?\s*&|,?\s+and)\s+/);
    return parts.map((part) => (part.endsWith(".") ? part : `${part}.`));
}

export const __all__ = ["parseApaCitation"];
